test/inductor/test_mkldnn_pattern_matcher.py

```diff
@@ -1145,7 +1145,7 @@ def matcher_check_fn():
                 counters["inductor"]["qconv2d_weight_prepack_matcher_nodes"], 12
             )
             # 2. Qconv2d Binary fusion in post-grad fusion pass * 1
-            #    [qconv2d_pointwise_default_1, convert_element_type_5, sub_2, mul_5, add_3, mul_6, round_4, add_4,
+            #    [qconv2d_pointwise_default_1, convert_element_type_5, sub_2, mul_5, add_3, mul_6, round_4, add_4 (optional),
             #     clamp_min_3, clamp_max_3, convert_element_type_6]
             self.assertEqual(counters["inductor"]["qconv2d_binary_matcher_count"], 1)
```

Comment:
A question for @jerryzh168. It seems like the current implementation of quantization is:
1. dynamo captures `torch.ops.quantized_decomposed.foo`
2. AOTDispatch hits a `CompositeExplicitAutograd` implementation which decomposes `foo` into `aten` ops
3. Inductor's pattern matcher matches those aten ops to recover the original meaning of `quantized_decomposed.foo`.

My question is why not keep `quantized.foo` in the AOTDispatch graph? This would make pattern matching much simpler and ensure that joint graph passes don't mess up the pattern before it can be matched in the post-grad passes. If the concern is not wanting to write lowerings, then you should still be able to decompose any quantization ops that remain after the quantization pattern matcher is finished.

Reply:
Or I suppose maybe you want all core aten ops for export's sake. In which case I think the best thing would be to have a set of "un-decompose" patterns that runs on the joint graph to undo such decompositions before any other passes run.

Reply:
cc @leslie-fang-intel

Reply:
I'm not exactly sure, maybe these are decomposed for performance optimizations, maybe @leslie-fang-intel or @Xia-Weiwen knows

Reply:
Hi @peterbell10 @jerryzh168, thanks for your comments. Also cc @jgong5 for this discussion.

- Yean, currently we do the `torch.ops.quantized_decomposed.quant/dequant` decomposition during the `AOTAutograd` which is before `pattern matcher phase`. I suppose we can move it to the Inductor `lowering phase` and removing corresponding decomposition. But it may require some efforts to ensure no bc-breaking.
- Another question, I think Quantization is just one of the specific cases which has been broken with the changes in https://github.com/pytorch/pytorch/pull/123445. Some other operators which has been decomposed with `add/mul` may also have potential vulnerabilities to this change. Can we move this pass after pattern matcher to avoid the breaking of existing patterns?

Reply:
> Can we move this pass after pattern matcher to avoid the breaking of existing patterns?

This pass is done during the joint graph pass which makes sense to me as the more we simplify the graph before partitioning, potentially the better partition we can get.  On the other hand the quantization patterns are registered post-grad which will always be after the joint graph phase.

One solution might be to move quantization passes into the joint graph phase, and move this pass to the end of the joint graph passes. That would allow us to match quantization patterns on the raw decomposed graph while still partitioning the simplified graph.

Does that sound acceptable?

Reply:
So looks like we may have 2 options now. @peterbell10 please correct me if any misunderstanding.

- Option 1: Don't do the `quant/dequant` decomposition and left the optimization in the lowering phase. In this way we have see these operators (un-decomposed) in the quantization pattern matcher phase. This option makes the quantization pattern clearer but may require some effort to refactor and ensure no bc-breaking.
- Option 2: As @peterbell10 mentioned above as move the quantization passes into the joint graph phase, and move `remove_no_ops` pass to the end of the joint graph passes. The change effort should be less than option 1.

@jgong5 any suggestions?

Reply:
@jgong5 ping